Solving DC circuits using numpy

a-elhag opened the pull request:
First off I want to thank @melissawm and the numpy team for all the encouragement and motivation! 

## Reason for writing tutorial:

If you google `how to solve circuits using matlab`, a plethora of results show up. At a cursory glance I counted: 3 books written on the topic, numerous youtube videos and countless articles. On the other hand, googling `how to solve circuits using numpy` I found only one [webpage](https://scipython.com/book/chapter-6-numpy/examples/mesh-analysis-of-a-electrical-network/) on the matter. In it they presented a grand total of 3 lines of code and in it they used the soon-to-be deprecated `matrix` class and `inverted` the matrix to solve it:
 
![NumpyTut](https://user-images.githubusercontent.com/48454648/97818161-e510dd00-1c6e-11eb-83a4-dd4d6995085a.png)

### Numpy modules used in the tutorial:
`np.array`, `np.linspace`, `np.argmax`, `np.linalg.inv`, `np.matmul`, @`, `np.linalg.solve`, `np.linalg.det`

---

##  Issues I faced:

1. The file is formatted with `MyST` and compiled perfectly with `sphinx`, but I had difficulties converting it to `.ipynb`. 
2. Didn't know how to show the outputs of the code so I left them blank except for the plot
3. Don't know how to reference images taken from the internet. For that reason, I drew the circuits by hand although I am looking into open source circuit drawing software so that anyone can add on to the tutorial in the future with the same consistency. 

## Pikachu

Added a couple of pictures of the lovable Pokemon but I don't know if it is suitable:
1. The google docs recommended not using local pop cultural references so that the tutorial could be understood worldwide. To that regard, I used to watch the show in Arabic and the [wiki](https://bulbapedia.bulbagarden.net/wiki/Dub) has it dubbed to +40 languages. I think you would be hard pressed to find someone who doesn't know pikachu
2. On the other hand, the google docs recommended not including 'frivolous` material which this probably falls under
3. But on the other "other" hand, the [numpy tutorial style guide](github.com/numpy/numpy-tutorials/blob/master/content/tutorial-style-guide.ipynbtu) advises to add an illustration below the title even if it is decorative. 

It's a very simple fix to remove, so do let me know to that regard.

melissawm commented:
Thank you, @a-elhag ! I'll take a look as soon as possible, please let me know if you have any questions in the meantime.

a-elhag commented:
So I updated it with `dc2.md`. It has all sphinx directives and python outputs removed.


I kept the older one for reference (can easily remove it though).

melissawm commented on content/electrical/dc.md:
````diff
@@ -0,0 +1,543 @@
+x = np.linalg.solve(A, b)
+print(x)
+```
+
````
Again, it would be nice to add some commentary on what the user sees here. What does `x` represent? What values are we supposed to expect?

melissawm commented on content/electrical/dc.md:
```diff
@@ -0,0 +1,543 @@
+
+### In practice...
+
+The python code written in `Part 1b` can indeed be written without any
+`for-loops` (use `range` to set the value of $V$)
```
```suggestion
- The code written in `Part 1b` can indeed be written without any `for-loops` (use `range` to set the value of $V$).
- There are many occasions in which the `scipy.linalg` module is preferrable over the `numpy.linalg` module. For more information on this, check the [scipy.linalg Reference](https://docs.scipy.org/doc/scipy/reference/tutorial/linalg.html).
```

melissawm reviewed (changes requested):
Hi @a-elhag! Thank you for writing this, I think this is a nice tutorial for someone looking to get started, especially if they have such a technical background. I have left a number of comments; please don't be discouraged - most of them are pretty small and simple :) In many cases, I'm using the conventions we have been using for the other tutorials, because this helps keeps the content consistent across the repo and makes it easier for users to know what to expect from each document.

Some specific comments:
- I LOVE Pikachu, but I'm not sure if we should include the pictures here. As far as I know, these are copyrighted and even if it could be argued that this is fair use, I think it's safer to drop it.
- In general, we tend to avoid the `admonition` role. It may clutter the page, so we should leave that for really important points that the reader can't miss. Some of my suggestions were moving the content in the admonitions to the main text.
- I added a bunch of reference links. Can you check that I added the appropriate ones?
- I'm suggesting we move the proposed exercises to the "on your own" section - I think that's more appropriate and makes the tutorial flow better because the user doesn't feel like they need to stop and do the exercises in the middle of it.
- I suggested this in a few comments but there is a last example where I didn't: whenever you choose specific values to run an example, make it clear to the user that they were chosen arbitrarily. This helps the user understand that they can play with those values and get different results without changing the main point of the analysis.

I think that's everything. Again, thank you for working on this and for considering these suggestions. If you have any questions or comments, let me know. Cheers!

a-elhag commented:
Hey @melissawm,

Thanks for taking the time to look through it! I'm here to learn and the more comments the better :)

I'm going through the suggestions now, will post an update when I'm done.

a-elhag commented on content/electrical/dc.md:
````diff
@@ -0,0 +1,543 @@
+plt.plot(Rload, Pload)
+```
+
+```{figure} imgs/plot2_better.png
+---
+width: 400px
+---
+```
````
I left both `dc.md` and `dc2.md` up. The later is the updated one and it doesn't have these. I'll delete the former so it doesn't cause confusion.

a-elhag commented on content/electrical/dc.md:
````diff
@@ -0,0 +1,543 @@
+Figure 3: Maximum Power Transfer Circuit
+```
+
+Electrical engineers claim that maximum power transfer will occur when Rload ==
+Rth. Write python code using NumPy arrays to test this.
````
I was writing on the assumption that people reading are electrical engineers. But it's a great point that more people would probably want to read it, I'll add a bit more background whenever I introduce a new concept.

a-elhag commented:
@melissawm 

I believe that I addressed each of the concerns, but do let me know if I missed anything. A summary of what was updated:

1. Deleted the markdown file with sphinx directives and kept only the one that was linked to the .ipynb file
2. Tried to rewrite points to make it clearer to a broader audience
3. Checked the reference links and they are appropriate
4. Emphasised that the numbers chosen were arbitrary. 


PS:
Although it saddens my heart, I removed all pictures of the lovable Pokemon. The copyright issue is a concern and the tonality might be too informal.

MarsBarLee commented:
Hey Al,

I have a suggestion- what do you think about updating your hand-drawn circuit diagrams? I have an example below of your first diagram.

![circuit1 again](https://user-images.githubusercontent.com/46167686/99984259-d489e900-2d7a-11eb-8119-d5e6aaa27a49.png)

Not a big difference for the first circuit, but I can help with the legibility and layout of the last few diagrams too.

I used a combination of an [online circuit diagram maker](https://www.circuit-diagram.org/) and Photoshop.

I could update the rest too. Let me know if you would like this. :)

a-elhag commented:
Yeah, thanks a lot @MarsBarLee! If you have any questions do let me know.

MarsBarLee commented:
Hi Al,

These are my updates to the diagrams. Let me know what you think!
![circuit1 edit](https://user-images.githubusercontent.com/46167686/101217274-12c2ba80-364f-11eb-91d6-a606791f0275.png)
![circuit2 edit](https://user-images.githubusercontent.com/46167686/101217275-135b5100-364f-11eb-9d16-d8eeb81fcbde.png)
![circuit3 edit](https://user-images.githubusercontent.com/46167686/101217276-135b5100-364f-11eb-9642-a4e8f2bf60e8.png)
![circuit4 edit](https://user-images.githubusercontent.com/46167686/101217278-135b5100-364f-11eb-92bb-6fa5358e143d.png)
![circuit5 edit](https://user-images.githubusercontent.com/46167686/101217279-135b5100-364f-11eb-879a-2e042a1016e4.png)
![circuit6 edit](https://user-images.githubusercontent.com/46167686/101217280-13f3e780-364f-11eb-84d9-27332e96f814.png)

a-elhag commented:
@MarsBarLee these look amazing! If you want to you can merge them in to this PR (is that possible?) or I can just do it myself. Thanks for these, mine looked like chicken scratch :smile:

MarsBarLee commented:
I looked up how to modify other people's PR (looks like creator would have to check 'Allow edits from maintainers' and I would need to push an edit?), but I think it would be easier for both of us if you did.

And no problem, happy to help!

a-elhag commented:
@melissawm I rebased the commits and updated the figures. Do let me know if there is anything more that I need to do.

And @MarsBarLee, yeah I didn't know if the process was simple or hard.

melissawm commented:
Hi @a-elhag ! Finally getting around to this - thanks for your patience!

Before doing a thorough review, I think we can get rid of the `.ipynb` file as we'll do everything via `.md`. Also, the `dc.md` and other files should be at the `content/` folder, not under another folder. We're still not 100% on the right organization of files in the repo (see #45 ) but for now everything lives under content.

melissawm commented on content/electrical/dc.md:
```diff
@@ -0,0 +1,452 @@
+---
+jupyter:
```
I don't know why this line is here, it doesn't show up for the other tutorials. Please check that this header is the same as for the other documents, so we don't have problems with jupytext and mystnb/sphinx.

melissawm reviewed (changes requested):
Done! You'll probably have to change the links to the images as well because of the folder structure and markdown format. To test if your document renders successfully, go to the `site` folder and run `make html` (you will have to have sphinx installed and maybe some other dependencies, you can check for those at the `requirements.txt` file.) You should also add this entry to the TOC at `site/index.md` for sphinx to find it. 

Thanks again, @a-elhag !

melissawm commented:
Hi @a-elhag - do you have any plans to continue work on this PR?

melissawm commented:
@a-elhag hello! If you don't have plans to update this tutorial I might take this over just to get the finishing touches and merge. Please let me know what you prefer. Cheers!
